modin/core/io/column_stores/parquet_dispatcher.py

```diff
@@ -63,73 +120,211 @@ def _unstrip_protocol(protocol, path):
                     return path
             return f"{protos[0]}://{path}"
 
-        if isinstance(path, AbstractBufferedFile):
-            return path.fs, [path]
-        fs = get_filesystem_class(split_protocol(path)[0])(**storage_options)
-        dataset = cls._get_dataset(path, storage_options)
+        if isinstance(self.path, AbstractBufferedFile):
+            return self.path.fs, [self.path]
+        fs = get_filesystem_class(split_protocol(self.path)[0])(**self.storage_options)
         # version.parse() is expensive, so we can split this into two separate loops
         if version.parse(fsspec.__version__) < version.parse("2022.5.0"):
-            files = [_unstrip_protocol(fs.protocol, fpath) for fpath in dataset.files]
+            files = [_unstrip_protocol(fs.protocol, fpath) for fpath in files]
         else:
-            files = [fs.unstrip_protocol(fpath) for fpath in dataset.files]
+            files = [fs.unstrip_protocol(fpath) for fpath in files]
 
         return fs, files
 
-    @classmethod
-    def _get_dataset(cls, path, storage_options):
-        """
-        Retrieve parquet dataset for given path.
 
-        Parameters
-        ----------
-        path : str, path object or file-like object
-            Path to dataset.
-        storage_options : dict
-            Parameters for specific storage engine.
+@_inherit_docstrings(Dataset)
+class PyArrowDataset(Dataset):
+    def __init__(self, path, storage_options):
+        self.path = path
+        self.storage_options = storage_options
+        self.dataset = self._init_dataset()
+        self._row_groups_per_file = None
+        self.filesystem, self._files = self._get_files(self.dataset.files)
 
-        Returns
-        -------
-        dataset : ParquetDataset
-            Dataset object of given Parquet file-path.
-        """
+    def _init_dataset(self):  # noqa: GL08
         from pyarrow.parquet import ParquetDataset
 
-        fs, fs_path = cls._get_fs_and_fs_path(path, storage_options)
+        fs, fs_path = self._get_fs_and_fs_path()
```

Comment:
@mvashishtha there are a few times where I repeat getting the file system from the path and I'm trying to reduce the amount of code duplication (see `_get_files`). I was a little lazy and kind of hacked a solution initially but I want to encapsulate these properly. I was wondering if you had any ideas?

Reply:
I see one `get_filesystem_class` call in each parquet class, and each class calls the same `_get_fs_and_fs_path`, which calls `url_to_fs`. I think this level of repetition is ok. Alternatively we could have each class internally call `self.fs` and `self.fs_path`, both of which compute the filesystem and `fs_path` and cache them.

Reply:
Oh I see, that could work,  so make both `fs` and `fs_path` as properties and cache them after the first call?

Reply:
Yeah, whenever someone calls `self.fs` or `self.fs_path`, compute both and cache them. But I think this is a minor improvement and the code as is is fine with me.

Reply:
Oh I actually really like that. That's clever, don't know why I didn't think of that 😃 . I think it's worth doing either way since I expect to get other comments for other style fixes.